File: libraries/react-native/ios/VoidhashEngineCore.swift
import Foundation
import Voidhash
import VoidhashCore

/// The Nitro-free heart of ``HybridVoidhashEngine``: owns the embedded bare-native client and
/// performs every data-plane operation on it.
///
/// Kept separate from the hybrid so the React Native Swift package can unit test the engine
/// without React Native or Nitro; the hybrid only wraps these calls into Nitro promises.
final class VoidhashEngineCore: @unchecked Sendable {
    /// Builds the embedded client; injected by tests to control the client's dependencies.
    typealias ClientFactory = @Sendable (_ publishableKey: String, _ options: VoidhashOptions) ->
        VoidhashClient

    struct EngineOptions: Decodable {
        var baseUrl: String?
        var ingestUrl: String?
        var debug: Bool?
        var enabled: Bool?
        var readOnly: Bool?
        var dev: Bool?
    }

    private struct IdentifyBody: Decodable {
        var distinctId: String
        var email: String?
        var name: String?
    }

    private struct AttributesBody: Decodable {
        var email: String?
        var name: String?
        var traits: [String: JSONValue]?
    }

    private struct SyncRequestEnvelope: Decodable {
        var distinctId: String
        var request: SdkSyncTransactionBody
    }

    private struct DevelopmentRequestEnvelope: Decodable {
        var distinctId: String
        var request: SdkDevelopmentPurchaseBody
    }

    static let notConfigured = VoidhashStoreError(
        code: "CONFIGURATION_MISSING", message: "VoidhashEngine is not configured")

    private let lock = NSLock()
    private let makeClient: ClientFactory
    private var client: VoidhashClient?

    init(
        makeClient: @escaping ClientFactory = { publishableKey, options in
            VoidhashClient(publishableKey: publishableKey, options: options)
        }
    ) {
        self.makeClient = makeClient
    }

    /// Maps the JSON the TypeScript client sends onto ``VoidhashOptions``. Absent keys keep the
    /// bare SDK's defaults, so a missing `readOnly` means observer mode.
    static func decodeOptions(_ optionsJson: String) throws -> VoidhashOptions {
        let options = try JSONDecoder().decode(EngineOptions.self, from: Data(optionsJson.utf8))
        var voidhashOptions = VoidhashOptions()
        if let baseUrl = options.baseUrl, let url = URL(string: baseUrl) {
            voidhashOptions.baseUrl = url
        }
        if let ingestUrl = options.ingestUrl {
            voidhashOptions.ingestUrl = URL(string: ingestUrl)
        }
        if let debug = options.debug {
            voidhashOptions.debug = debug
        }
        if let enabled = options.enabled {
            voidhashOptions.enabled = enabled
        }
        if let readOnly = options.readOnly {
            voidhashOptions.readOnly = readOnly
        }
        if let dev = options.dev {
            voidhashOptions.dev = dev
        }
        return voidhashOptions
    }

    /// Creates the embedded client. The client is never started: the engine is data-plane only
    /// and must not open a second store observer next to the one the JS layer owns.
    func configure(publishableKey: String, optionsJson: String) throws {
        let options = try Self.decodeOptions(optionsJson)
        let client = makeClient(publishableKey, options)
        lock.withLock { self.client = client }
    }

    var isConfigured: Bool {
        return lock.withLock { client != nil }
    }

    /// Whether the embedded client runs in observer mode; `nil` before ``configure``.
    var isReadOnly: Bool? {
        return lock.withLock { client }?.isReadOnly
    }

    /// Mirrors the JS observer-mode decision into the embedded client, which owns the flag the
    /// `x-observer-mode` header is built from.
    func setReadOnly(_ readOnly: Bool) throws {
        try requireClient().setReadOnly(readOnly)
    }

    func fetchSchema(distinctId: String) async throws -> String {
        let schema = try await requireClient().fetchSchema(distinctId: distinctId)
        return try Self.encode(schema)
    }

    func fetchPerson(distinctId: String) async throws -> String {
        guard let person = try await requireClient().fetchPerson(distinctId: distinctId) else {
            return "null"
        }
        return try Self.encode(person)
    }

    func identify(distinctId: String, bodyJson: String) async throws -> String {
        let body = try JSONDecoder().decode(IdentifyBody.self, from: Data(bodyJson.utf8))
        let person = try await requireClient().identifyPerson(
            distinctId: distinctId,
            externalUserId: body.distinctId,
            email: body.email,
            name: body.name
        )
        return try Self.encode(person)
    }

    func setPersonAttributes(distinctId: String, attributesJson: String) async throws -> String {
        let body = try JSONDecoder().decode(AttributesBody.self, from: Data(attributesJson.utf8))
        var traits: [String: JSONValue] = body.traits ?? [:]
        if let email = body.email {
            traits["email"] = .string(email)
        }
        if let name = body.name {
            traits["name"] = .string(name)
        }
        let person = try await requireClient().setPersonTraits(
            distinctId: distinctId, traits: traits)
        return try Self.encode(person)
    }

    func evaluateFlags(distinctId: String, flagKeysJson: String) async throws -> String {
        let keys = try JSONDecoder().decode([String].self, from: Data(flagKeysJson.utf8))
        let flags = try await requireClient().evaluateFeatureFlags(
            distinctId: distinctId, keys: keys.isEmpty ? nil : keys)
        return try Self.encode(flags)
    }

    func resolvePaywall(distinctId: String, locationSlug: String) async throws -> String {
        guard
            let paywall = try await requireClient().resolvePaywallConfig(
                distinctId: distinctId, locationSlug: locationSlug)
        else {
            return "null"
        }
        return try Self.encode(paywall)
    }

    func syncTransaction(distinctId: String, requestJson: String) async throws -> Bool {
        let envelope = try JSONDecoder().decode(
            SyncRequestEnvelope.self, from: Data(requestJson.utf8))
        return try await requireClient().syncStoreTransaction(
            envelope.request,
            distinctId: envelope.distinctId.isEmpty ? distinctId : envelope.distinctId
        )
    }

    func developmentPurchase(distinctId: String, requestJson: String) async throws -> Bool {
        let envelope = try JSONDecoder().decode(
            DevelopmentRequestEnvelope.self, from: Data(requestJson.utf8))
        try await requireClient().recordDevelopmentPurchase(
            envelope.request,
            distinctId: envelope.distinctId.isEmpty ? distinctId : envelope.distinctId
        )
        return true
    }

    func injectInternalSchema(schemaJson: String) async throws {
        let schema = try JSONDecoder().decode(RuntimeSchema.self, from: Data(schemaJson.utf8))
        try await requireClient().injectInternalSchema(schema)
    }

    private func requireClient() throws -> VoidhashClient {
        guard let client = lock.withLock({ client }) else {
            throw Self.notConfigured
        }
        return client
    }

    private static func encode<Value: Encodable>(_ value: Value) throws -> String {
        let data = try JSONEncoder().encode(value)
        return String(data: data, encoding: .utf8) ?? "null"
    }
}

File: libraries/react-native/android/src/main/java/com/margelo/nitro/voidhash/engine/VoidhashEngineCore.kt
package com.margelo.nitro.voidhash.engine

import android.content.Context
import com.voidhash.sdk.ScreenTrackingOptions
import com.voidhash.sdk.VOIDHASH_DEFAULT_BASE_URL
import com.voidhash.sdk.Voidhash
import com.voidhash.sdk.VoidhashClient
import com.voidhash.sdk.VoidhashOptions
import com.voidhash.sdk.api.DevelopmentPurchaseRequest
import com.voidhash.sdk.api.SdkHeaders
import com.voidhash.sdk.api.SyncTransactionRequest
import com.voidhash.sdk.api.VoidhashApiClient
import com.voidhash.sdk.platform.PlatformInfo
import org.json.JSONArray
import org.json.JSONObject

private fun JSONObject.optStringOrNull(key: String): String? {
    if (!has(key) || isNull(key)) return null
    return optString(key)
}

/**
 * The Nitro-free heart of `HybridVoidhashEngine`: owns the embedded bare-native client and
 * performs every data-plane operation on it.
 *
 * Kept separate from the hybrid so it can be unit tested on the JVM without React Native or
 * Nitro; the hybrid only wraps these calls into Nitro promises. The factories exist for tests.
 */
class VoidhashEngineCore(
    private val configureClient: (Context, String, VoidhashOptions) -> VoidhashClient =
        { context, publishableKey, options -> Voidhash.configure(context, publishableKey, options) },
    private val platformInfoProvider: (Context) -> PlatformInfo = PlatformInfo::fromContext,
    private val apiClientFactory: (String, SdkHeaders) -> VoidhashApiClient =
        { baseUrl, headers -> VoidhashApiClient(baseUrl, headers) },
) {
    /**
     * The options JSON the TypeScript client sends. Absent keys keep the bare SDK's defaults,
     * so a missing `readOnly` means observer mode.
     *
     * Automatic native screens default to off: the JS layer owns screen tracking and passes
     * `screenTracking.automatic` explicitly. The `$app_*` lifecycle events are always emitted
     * by JS, so the embedded client never captures them.
     */
    class Options(optionsJson: String) {
        private val json = JSONObject(optionsJson)

        val baseUrl: String = json.optStringOrNull("baseUrl") ?: VOIDHASH_DEFAULT_BASE_URL
        val ingestUrl: String? = json.optStringOrNull("ingestUrl")
        val debug: Boolean = json.optBoolean("debug", false)
        val enabled: Boolean = json.optBoolean("enabled", true)
        val readOnly: Boolean = json.optBoolean("readOnly", true)
        val dev: Boolean = json.optBoolean("dev", false)
        val automaticScreenTracking: Boolean =
            json.optJSONObject("screenTracking")?.optBoolean("automatic", false) ?: false

        fun toVoidhashOptions(): VoidhashOptions = VoidhashOptions(
            baseUrl = baseUrl,
            ingestUrl = ingestUrl,
            debug = debug,
            enabled = enabled,
            readOnly = readOnly,
            dev = dev,
            screenTracking = ScreenTrackingOptions(automatic = automaticScreenTracking),
            automaticLifecycleEvents = false,
        )
    }

    @Volatile
    private var apiClient: VoidhashApiClient? = null

    @Volatile
    private var client: VoidhashClient? = null

    /** The header builder the engine's requests use; exposed for the tests. */
    @Volatile
    internal var headers: SdkHeaders? = null
        private set

    val isConfigured: Boolean get() = client != null

    /** Whether the embedded client runs in observer mode; `null` before [configure]. */
    val isReadOnly: Boolean? get() = client?.isReadOnly

    /**
     * Creates the embedded client. The client is never initialized: the engine is data-plane
     * only and must not open a billing connection next to the observer the JS layer owns.
     */
    fun configure(context: Context, publishableKey: String, optionsJson: String) {
        val platform = platformInfoProvider(context)
        val options = Options(optionsJson)
        val configuredClient = configureClient(context, publishableKey, options.toVoidhashOptions())
        // The client owns the observer flag, so the header can never drift from what the
        // client — and through `setReadOnly`, the JS layer — decided.
        val sdkHeaders = SdkHeaders(
            publishableKey = publishableKey,
            platform = platform,
            readOnlyProvider = { configuredClient.isReadOnly },
            debugProvider = { options.debug },
            environmentProvider = {
                if (options.dev && platform.isDebugBuild) "development" else "production"
            },
        )
        client = configuredClient
        headers = sdkHeaders
        apiClient = apiClientFactory(options.baseUrl, sdkHeaders)
    }

    /** Mirrors the JS observer-mode decision into the embedded client. */
    fun setReadOnly(readOnly: Boolean) {
        requireClient().setReadOnly(readOnly)
    }

    suspend fun fetchSchema(distinctId: String): String =
        requireApiClient().getSchema(distinctId).toString()

    suspend fun fetchPerson(distinctId: String): String =
        requireApiClient().getPerson(distinctId)?.raw?.toString() ?: "null"

    suspend fun identify(distinctId: String, bodyJson: String): String {
        val body = JSONObject(bodyJson)
        val person = requireApiClient().identify(
            distinctId,
            body.getString("distinctId"),
            body.optStringOrNull("email"),
            body.optStringOrNull("name"),
        )
        return person.raw.toString()
    }

    suspend fun setPersonAttributes(distinctId: String, attributesJson: String): String {
        val body = JSONObject(attributesJson)
        val traits = body.optJSONObject("traits") ?: JSONObject()
        body.optStringOrNull("email")?.let { traits.put("email", it) }
        body.optStringOrNull("name")?.let { traits.put("name", it) }

        val attributes = LinkedHashMap<String, Any?>()
        for (key in traits.keys()) {
            attributes[key] = traits.get(key).takeUnless { it == JSONObject.NULL }
        }
        val apiClient = requireApiClient()
        apiClient.setPersonAttributes(distinctId, attributes)
        return apiClient.getPerson(distinctId)?.raw?.toString() ?: "null"
    }

    suspend fun evaluateFlags(distinctId: String, flagKeysJson: String): String {
        val keys = JSONArray(flagKeysJson).let { array ->
            (0 until array.length()).map { array.getString(it) }
        }
        val flags = requireApiClient().evaluateFlags(distinctId, keys)
        val response = JSONArray()
        for (flag in flags) {
            response.put(
                JSONObject()
                    .put("key", flag.key)
                    .put("enabled", flag.enabled)
                    .put("variantKey", flag.variantKey ?: JSONObject.NULL),
            )
        }
        return response.toString()
    }

    suspend fun resolvePaywall(distinctId: String, locationSlug: String): String =
        requireApiClient().resolvePaywallRaw(distinctId, locationSlug)?.toString() ?: "null"

    suspend fun syncTransaction(distinctId: String, requestJson: String): Boolean {
        val request = JSONObject(requestJson).getJSONObject("request")
        return requireApiClient().syncTransaction(
            distinctId,
            SyncTransactionRequest(
                appAccountToken = request.optStringOrNull("appAccountToken"),
                providerProductId = request.optStringOrNull("providerProductId") ?: "",
                productSlug = request.getString("productSlug"),
                purchaseDate = request.getDouble("purchaseDate"),
                purchaseToken = request.optStringOrNull("purchaseToken") ?: "",
                quantity = request.optInt("quantity", 1),
                receipt = request.optStringOrNull("receipt"),
                transactionId = request.getString("transactionId"),
            ),
        )
    }

    suspend fun developmentPurchase(distinctId: String, requestJson: String): Boolean {
        val request = JSONObject(requestJson).getJSONObject("request")
        requireApiClient().developmentPurchase(
            distinctId,
            DevelopmentPurchaseRequest(
                devTransactionId = request.getString("devTransactionId"),
                productSlug = request.getString("productSlug"),
                purchaseDate = request.getDouble("purchaseDate"),
                quantity = request.optInt("quantity", 1),
            ),
        )
        return true
    }

    fun injectInternalSchema(schemaJson: String) {
        requireClient().injectInternalSchema(
            com.voidhash.sdk.schema.RuntimeSchema.fromJson(JSONObject(schemaJson)),
        )
    }

    private fun requireClient(): VoidhashClient =
        client ?: throw IllegalStateException(NOT_CONFIGURED_MESSAGE)

    private fun requireApiClient(): VoidhashApiClient =
        apiClient ?: throw IllegalStateException(NOT_CONFIGURED_MESSAGE)

    companion object {
        const val NOT_CONFIGURED_MESSAGE = "CONFIGURATION_MISSING: VoidhashEngine is not configured"
    }
}
